### ISSUE TITLE
Comparison to Pweave

### ISSUE THREAD

**Issue by piccolbo** [danijar/handout#12]
Just a quick note to ask if you are aware of pweave project, which, albeit without an active maintainer at the moment, does exactly the same. Supports markdown with python snippets, modeled after Rstudio Rmd format, or python with special comments as in this project. Supports matplotlib as well bokeh graphics (I have patches for altair, also web based and with interactions).

**Comment by epogrebnyak** [comment]
[Pweave](https://github.com/mpastell/Pweave) is certainly a worthwhile project, but has a different ideology: the source is not a runnable python script, it is a markdown file with embedded python code.

**Comment by danijar** [comment]
Thanks for bringing this to my attention, @piccolbo. I'm aware of the project. @epogrebnyak summarized one of the key motivations for Handout very well.

**Comment by epogrebnyak** [comment]
Just for the sake of completeness there other two python projects that embed code inside markdown, similar to pweave:
- <https://github.com/pystitch/stitch>
- <https://github.com/jankatins/knitpy>

Both seem not in active development. I think these tools are more developed and widely used in R, so these users stay with original [knitr](http://yihui.name/knitr/) and [rmarkdown](http://rmarkdown.rstudio.com/).

**Comment by piccolbo** [comment]
As I said pweave supports working on commented py files.  That is specifically a commonality between the two projects. All the tests of my project altair_recipes, for instance, are valid python files run by pytest, but are also processed with pweave to create additional examples.

**Comment by epogrebnyak** [comment]
@piccolbo - is your project on github? If you can share a link or a small part of code here, maybe we can put an extra example, contrasting the use of `pweave` and `handout`.  Or perhaps, interchangeability of these two tools?

**Comment by piccolbo** [comment]
Not too pleased there are so many similar projects not being developed. As I said, pweave development also slowed down. I hope you guys can stick with it and see it take off. But I don't think you should reject supporting other formats , say Pmd,  off the bat. It's still just parsing, executing code, capturing results, etc. Most of the process is shared whichever syntax  you start from. As to you request, you bet: [python code](https://github.com/piccolbo/altair_recipes/tree/master/tests),  [Makefile](https://github.com/piccolbo/altair_recipes/blob/master/Makefile) (check the docs: target) and [doc output](https://altair-recipes.readthedocs.io/en/latest/examples.html)

**Comment by piccolbo** [comment]
One choice in pweave is the use of special comments that start with %' or $%% so that you can still have non-markdown regular comments. I think this is in keeping with a tradition of embedded documentation systems in other languages if I can remember. If I understand correctly handout processes any triple quote comment. So for instance it will hit docstrings as well. Still thinking through the pros and cons. It sure does look cleaner in the source but you lose one deg of flexibility. Does it matter?

**Comment by epogrebnyak** [comment]
>  So for instance it will hit docstrings as well.

Good point! One could hope the docstrings can be left intact via `# handout: exclude` directive, but did not work for me in `example.py`

> sure does look cleaner in the source

Indeed, very traceable code @danijar made!

**Comment by danijar** [comment]
@piccolbo Thanks for bringing up docstrings. I've opened #15 to discuss an appropriate syntax.

Regarding converting scripts to reports without running them, this is not a goal of Handout.

There are a couple of projects mentioned above that go this route. However, one benefit of Handout is that you can create and update the report while the script is still running. For example, this is useful for machine learning to visualize or print statistics of an algorithm while it's training.

In other words, Handout will stay a Python package that is imported in a script that can be run normally, e.g. as `python3 script.py`.
